documentapi/src/vespa/documentapi/messagebus/routable_factories_8.cpp

```diff
@@ -0,0 +1,883 @@
+// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#include "routable_factories_8.h"
+#include <vespa/document/bucket/bucketidfactory.h>
+#include <vespa/document/fieldvalue/document.h>
+#include <vespa/document/select/parser.h>
+#include <vespa/document/update/documentupdate.h>
+#include <vespa/document/util/serializableexceptions.h>
+#include <vespa/documentapi/documentapi.h>
+#include <vespa/documentapi/messagebus/docapi_common.pb.h>
+#include <vespa/documentapi/messagebus/docapi_feed.pb.h>
+#include <vespa/documentapi/messagebus/docapi_visiting.pb.h>
+#include <vespa/documentapi/messagebus/docapi_inspect.pb.h>
+#include <vespa/vespalib/objects/nbostream.h>
+
+namespace documentapi::messagebus {
+
+namespace {
+
+// Protobuf codec helpers for common types
+
+void set_bucket_id(protobuf::BucketId& dest, const document::BucketId& src) {
+    dest.set_raw_id(src.getRawId());
+}
+
+document::BucketId get_bucket_id(const protobuf::BucketId& src) {
+    return document::BucketId(src.raw_id());
+}
+
+void set_document_id(protobuf::DocumentId& dest, const document::DocumentId& src) {
+    auto doc_id = src.toString();
+    dest.set_id(doc_id.data(), doc_id.size());
+}
+
+document::DocumentId get_document_id(const protobuf::DocumentId& src) {
+    return document::DocumentId(src.id());
+}
+
+// TODO DocumentAPI should be extended to use actual document::FieldSet enums instead of always passing strings.
+void set_raw_field_set(protobuf::FieldSet& dest, vespalib::stringref src) {
+    dest.set_spec(src.data(), src.size());
+}
+
+// Note: returns by ref
+vespalib::stringref get_raw_field_set(const protobuf::FieldSet& src) noexcept {
+    return src.spec();
+}
+
+void set_raw_selection(protobuf::DocumentSelection& dest, vespalib::stringref src) {
+    dest.set_selection(src.data(), src.size());
+}
+
+// Note: returns by ref
+vespalib::stringref get_raw_selection(const protobuf::DocumentSelection& src) noexcept {
+    return src.selection();
+}
+
+void set_bucket_space(protobuf::BucketSpace& dest, vespalib::stringref space_name) {
+    dest.set_name(space_name.data(), space_name.size());
+}
+
+// Note: returns by ref
+vespalib::stringref get_bucket_space(const protobuf::BucketSpace& src) noexcept {
+    return src.name();
+}
+
+void set_global_id(protobuf::GlobalId& dest, const document::GlobalId& src) {
+    char tmp[document::GlobalId::LENGTH];
+    memcpy(tmp, src.get(), document::GlobalId::LENGTH);
+    dest.set_raw_gid(tmp, document::GlobalId::LENGTH);
+}
+
+document::GlobalId get_global_id(const protobuf::GlobalId& src) {
+    if (src.raw_gid().size() != document::GlobalId::LENGTH) {
+        throw document::DeserializeException("Unexpected serialized protobuf GlobalId size");
```

Comment:
Consider adding actual and expected lengths to exception message.

Reply:
Done